### ISSUE TITLE
Support for DRM `syncobj` FDs in `VkSemaphoreGetFdInfoKHR`

### ISSUE THREAD

**Issue by colinmarc** [KhronosGroup/Vulkan-Docs#2473]
[DRM timeline sync objects](https://www.kernel.org/doc/html/v5.4/gpu/drm-mm.html?highlight=syncobj#drm-sync-objects) seem tailor-made for interop with Vulkan timeline semaphores. However, while it's possible to import these using [`vkSemaphoreFdKHR`](https://registry.khronos.org/VulkanSC/specs/1.0-extensions/man/html/vkGetSemaphoreFdKHR.html) and `VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT`, there are a few issues that make importing/exporting unergonomic:

 - When importing, the `SYNC_FD` type seems more correct, and it's not clear in the spec which should be used in this case (unless I missed it). However, `SYNC_FD` is restricted to temporary binary semaphore imports, and does not allow permanent timeline semaphore imports.
 - When exporting a semaphore as a FD using [`vkSemaphoreGetFdInfoKHR`](https://registry.khronos.org/VulkanSC/specs/1.0-extensions/man/html/VkSemaphoreGetFdInfoKHR.html), there's no way to distinguish what type of FD you're getting.

Should syncobjs be covered by an additional enum value in [`HandleTypeFlagBits`](https://registry.khronos.org/VulkanSC/specs/1.0-extensions/man/html/VkExternalSemaphoreHandleTypeFlagBits.html)? Or should the the spec/validation be relaxed to allow timeline semaphores to be imported/exported using `SYNC_FD`?

**Comment by oddhack** [comment]
@linyaa-kiwi this looks like an SI TSG issue.

**Comment by linyaa-kiwi** [comment]
cc @gfxstrand

**Comment by cubanismo** [comment]
Neither VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT (FD referring to an undefined (hence "opaque") object type with limited functionality requirements defined by Vulkan. Might refer to a DRM syncobj, might not) nor VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_SYNC_FD_BIT (Linux sync_file) correspond to DRM syncobj FDs. They would need a new handle type enum. There was some dithering in the past about whether DRM syncobj timelines were an object that should be surfaced by more application-visible APIs like Vulkan as opposed to treated as driver-internal constructs, but I think other parts of the ecosystem have embraced them enough that it probably makes sense to include them in Vulkan as well at this point.

**Comment by colinmarc** [comment]
@cubanismo importing a syncobj timeline from wayland as `VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT` actually just stopped working for me on NVIDIA 560/565. I swear it was working before. [Here](https://github.com/colinmarc/magic-mirror/blob/main/mm-server/src/vulkan/timeline.rs#L73-L81) is the code where I import it.

My DebugUtilsMessenger spits out the error `Failed to allocate semaphore device memory`. Is that intended to be working today or did I misunderstand something about the current support?

**Comment by cubanismo** [comment]
No, it's not expected to work. It would never have worked. The NV Linux driver doesn't use DRM syncobj (Or DRM at all) to implement timeline semaphores and never has, so at best it might have been silently failing before, but we just verified it's been returning the correct error message in this case while following up on a another bug report, so I'm doubtful that was happening.

**Comment by mahkoh** [comment]
On systems where VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT is syncobj you can detect this today by calling one of the DRM icotls that expect syncobj arguments. If so, you can use the syncobj as you would if there were an official extension, I believe.

**Comment by mahkoh** [comment]
I've opened #2692.

**Comment by cubanismo** [comment]
Just wanted to provide an update given there's been internal discussion and progress on this extension in the system integration TSG:

* Some Khronos members have reviewed the spec and CTS and consider them more or less complete.
* Besides the linked Mesa implementation, there is another Mesa-based implementation from a Khronos member, which means this proposal meets the minimum implementation requirements we maintain for publishing a multi-vendor (EXT) extension specification.

Hence, to the best of my knowledge this extension ticks all the boxes needed to start moving through the release process. I've taken steps to kick off the Khronos-internal aspects of that process.

**Comment by gfxstrand** [comment]
Two notes:

 1. As per the new Mesa process, the Mesa MR needs to be reviewed first: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/41400
 2. All Mesa implementations will be effectively the same implementation. I think EXT still makes sense as it will cover a LOT of drivers but if we're expecting multiple fundamentally different implementations, we're not going to get that.

**Comment by cubanismo** [comment]
> As per the new Mesa process, the Mesa MR needs to be reviewed first: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/41400

Does it? That process refers specifically to MESA-prefixed extensions if I'm reading the policy correctly. Regardless, it seems like a good idea.

> All Mesa implementations will be effectively the same implementation. I think EXT still makes sense as it will cover a LOT of drivers but if we're expecting multiple fundamentally different implementations, we're not going to get that.

I'll wait and see the code, but internal discussion suggested the other one may be more different than I had expected.

The alternative is waiting for the NV implementation, which will be quite different and hence there is some risk there, but I'd rather not hold this extension up if multiple other parties are ready to move forward. We could always do a maintenance/v2 extension in the unlikely case we find problems at the spec level.
